crates/motoko/tests/test_dynamic.rs

```diff
@@ -1,46 +1,71 @@
-use motoko::dynamic::Dynamic;
-use num_bigint::ToBigUint;
-use std::rc::Rc;
+use motoko::vm_types::{Interruption, Pointer};
+use motoko::{dynamic::Dynamic, value::Value_};
 
-use motoko::value::Text;
 use motoko::value::{DynamicValue, Value};
 
 #[test]
 fn dyn_struct() {
     #[derive(Clone, Debug, Hash, Default)]
     struct Struct {
-        pub map: im_rc::HashMap<Value, Rc<Value>>,
+        pub map: im_rc::HashMap<Value_, Value_>,
+        pub x: Option<Value_>,
     }
 
     impl Dynamic for Struct {
-        fn get_index(&self, index: &Value) -> Option<Rc<Value>> {
-            println!("Index: {:?}", index);
-            self.map.get(index).map(Clone::clone)
+        fn get_index(&self, index: &Value) -> motoko::dynamic::Result {
+            self.map
+                .get(index)
+                .map(Clone::clone)
+                .ok_or_else(|| Interruption::IndexOutOfBounds)
         }
 
-        fn get_field(&self, name: &str) -> Option<Rc<Value>> {
+        fn set_index(&mut self, key: Value_, value: Value_) -> motoko::dynamic::Result<()> {
+            self.map.insert(key, value);
+            Ok(())
+        }
+
+        fn get_field(&self, name: &str) -> motoko::dynamic::Result {
             match name {
-                "x" => Some(Rc::new(Value::Text(Text::from("expected")))),
-                _ => None,
+                "x" => Ok(self.x.clone().expect("`x` not defined")),
+                _ => Err(Interruption::UnboundIdentifer(name.to_string())),
             }
         }
+
+        fn set_field(&mut self, name: &str, value: Value_) -> motoko::dynamic::Result<()> {
+            match name {
+                "x" => {
+                    self.x = Some(value);
+                    Ok(())
+                }
+                _ => Err(Interruption::UnboundIdentifer(name.to_string())),
+            }
+        }
+
+        fn call(&self, _inst: &Option<motoko::ast::Inst>, args: Value_) -> motoko::dynamic::Result {
+            Ok(args)
+        }
     }
 
-    let expected = Rc::new(Value::Text(Text::from("expected")));
-    let value = Value::Dynamic(DynamicValue(Box::new({
-        let mut s = Struct {
-            map: Default::default(),
-        };
-        s.map
-            .insert(Value::Nat(5.to_biguint().unwrap()), expected.clone());
-        s
-    })));
+    let value = Value::Dynamic(DynamicValue(Box::new(Struct::default())));
 
     let mut core = motoko::vm_types::Core::empty();
-    core.env.insert("value".to_string(), value);
+
+    // TODO: use something like `core.alloc()`
+    let pointer = Pointer(12345);
```

Comment:
Yeah, we need a general allocation approach.

Relatedly, there's no "next pointer" field for `Core`, but there should be, and that should be the mechanism for choosing the next pointer name, not the size of the store.

Reply:
Done! 

Feel free to change things around, but this should work for now.